### ISSUE TITLE
definition of field

### ISSUE THREAD

**Issue by DanGrayson** [UniMath/UniMath#254]
Here is our definition of "field".

```
Definition invpair (X:monoid) (x:X) := Σ x':X, ( x' * x = 1 ) × ( x * x' = 1 ).
Definition multinvpair ( X : rig ) ( x : X ) := invpair ( rigmultmonoid X ) x .
Definition isafield (X:commrng) := isnonzerorng X  ×  ∀ x:X, multinvpair X x ⨿ (x = 0).
Definition fld := Σ X, isafield X.
```

Are there any examples worthy of being considered as fields that satisfy only the weaker condition 
that ` x != 0 -> multinvpair X x ` ?

The construction of the field of fractions for an integral domain, as a result of this, requires the
additional hypothesis that the ring has decidable equality.  I can imagine that being counterintuitive
when we start formalizing commutative algebra, and we want to associate a field to each
prime ideal of a commutative ring.

I'm tempted to weaken the definition to allow more natural developments, even if there are
no interesting examples.

**Comment by DanGrayson** [comment]
Here is a related comment.  In formalization of facts about integral domains
and fields, it is important to consider the multiplicative monoid of nonzero
elements.  But the proof that x!=0 is a proposition uses the axiom
`funextempty`, spoiling computability, so we will always want a proposition x≠0
to replace it that can be shown to be a proposition without using axioms.  When
equality is decidable, one is available, but in general, the development of the
real numbers with its apartness relation suggests that the right approach is
instead to seek a property x≠0 such that `(¬x≠0) <-> x=0`.  (With decidability,
one of these is always available, also.)

Currently the definition of "integral domain" is unsatisfying, as it uses
propositional truncation, thereby also introducing axioms:

```
Definition isintdom ( X : commrng ) :=
   dirprod ( isnonzerorng X )
           ( forall a1 a2 : X ,
	       paths ( a1 * a2 ) 0 -> hdisj ( eqset a1 0 ) ( eqset a2 0 ) ) .
```

Simpler looking would be:
```
Definition isintdom ( X : commrng ) := isnonzerorng X ×
	   ∀ x y, x ≠ 0 -> y ≠ 0 -> x*y ≠ 0.
```

If we adopt this approach, then our definition of "field" would be the same as
Catherine's definition of "constructive field", provided we add various
properties that say that the relation x≠y is compatible with addition and
multiplication.

Merging the two definitions seems worthwhile, at first glance.

**Comment by vladimirias** [comment]
> when we start formalizing commutative algebra, and we want to associate a field to each prime ideal of a commutative ring.
> 
I am not sure what is relevant here: http://www.cse.chalmers.se/~coquand/algebra.html for this question but I think it is worth looking at the papers on the space of valuations and on the one about points of curves.

**Comment by andrejbauer** [comment]
I am not an expert on this particular topic, but you seem to be undergoing the usual process of "constructivization" where you need to choose one among several classically equivalent formulations, and you're drawn towards the more constructive ones on the grounds that Coq likes them better. 

Regarding constructive algebra in general I would recommend looking at:

* [A course in constructive algebra](https://books.google.si/books?id=4W0QBwAAQBAJ)
* [Commutative algebra: Constructive methods](http://www.springer.com/us/book/9789401799430)

When you get to ideals you will likely consider *anti*ideals, which are "complements" of ideals phrased in terms of apartness. And there are also antisubgroups of groups with apartness. I am not sure, however, that going 100% constructive is the best strategy to follow.

**Comment by DanGrayson** [comment]
We have only one sensible choice for the constructive (not necessarily decidable) notion
of field, the one with an apartness relation.  My proposal amounts to a simple renaming
of the terms we currently use:
    field -> decidable field
    constructive field -> field
The point is to encourage as much development of algebra as possible without assuming
fields have decidable equality, so that as many facts as possible will apply to the reals
and the complexes.  Nothing will be lost thereby, and much may be gained.

**Comment by DanGrayson** [comment]
The definition of "field" in "A course in constructive Algebra", by Mines, Richman,
and Ruitenburg, is the one we call "constructive field".  So that supports my proposal.

**Comment by DanGrayson** [comment]
... and they call a field with decidable equality a "discrete field".  Actually they make
tightness optional (sort of silly) and call it a Heyting field if the inequality relation is tight.

**Comment by DanGrayson** [comment]
They also speak of polynomials over non-decidable rings, and point out that they may not
have a degree.

**Comment by mortberg** [comment]
We should check that from the definition of field we can prove that any field is a local ring. According to the book of Lombardi/Quitte the good constructive way to express that a ring R is local is: forall (x y : R), invertible (x + y) -> coprod (invertible x) (invertible y)

which is equivalent to: forall (x : R), coprod (invertible x) (invertible (1-x)) 

On page 491 of the english translation of the book they define fields as:

"A Heyting field, or simply a field, is by definition a local ring in which every
non-invertible element is null, in other words a local ring whose Jacobson
radical is reduced to 0.

In particular, a discrete field, therefore also the trivial ring, is a field. The
real numbers form a field that is not a discrete field"

I wonder why they include the information that the ring is local in the definition, maybe it is not provable from the other property? What about the other definitions of field we have discussed?

**Comment by DanGrayson** [comment]
That property, propositionally truncated, is the "cotransitivity" property of the apartness relation,
and is part of the definition of constructive field now.
Untruncated, it would be too strong for the real numbers to satisfy it, I guess.  Are you sure
that's what Lombardi meant?  @cathlelay, am I right?

For any prime ideal, not just for maximal ideals in local rings, it appears one should have a 
proposition whose negation amounts to membership in the ideal.  Then that proposition could
be used for forming the multiplicative set of element "not in" the ideal.

**Comment by DanGrayson** [comment]
Another point in favor of modifying the definition of "field" is so it will cover (complete) local fields, such as the p-adic numbers.  Their Galois theory is important, too.

**Comment by vladimirias** [comment]
I still think that “a field” and “an apartness field” are the best names with fld and apfld as UniMath abbreviations.

This thinking comes from a feeling that I had even back 20 years ago when I knew nothing about constructive mathematics - that fields such as, for example, the field of functions of a henselization of Z in p are somehow more algebraic objects than completions such as Z_p. In the constructive setting this feeling gets a precise expression - the field of functions of a henzelization, which has decidable equality, is a field while Z_p is an object of a different class that is located slightly away from algebra in the direction of where the study of ordered algebraic structures lies.

Vladimir.

**Comment by DanGrayson** [comment]
But don't you want our formalization of commutative algebra of polynomial rings
over fields and of Galois theory to apply to local fields without
having to assume excluded middle?  Class field theory includes a lot of Galois theory
of local fields, and with the wrong definition, the formalization of those areas will never
happen, because no one will have the energy to make a parallel development.

**Comment by vladimirias** [comment]
I think it is a good idea to develop it (our formalization of commutative algebra of polynomial rings) over apartness fields for as much as possible.

There is a function fld -> apfld isn’t there?

I think this is the first fact that needs to be firmly (formally) established. Following the definition of apfld there should be the definition of a function fld -> apfld

I would expect that for the axiom neg(isinv x) -> (x=0) to be of any use the axioms of an apartness relation should be required to hold for the relation isinv(x-y). Do they hold if (x=0)\coprod (isinv x) holds?

**Comment by DanGrayson** [comment]
Yes, there is a function fld -> apfld, because decidability of equality is part of the definition, and all the properties hold that you ask about.

But the natural reaction of a potential new contributor who is interested in developing some field theory will be to develop formalizations based on "fld", because of the name, which would be a mistake.  And our commutative algebra and Galois theory is not going to look nice to mathematicians if the name "fld" is nowhere to be seen and the screen is littered with the strange looking "apfld".

**Comment by DanGrayson** [comment]
If our field theory development is littered with concepts such as "finite apartness field extension", "separable apartness field extension", "normal apartness field extension", "Galois apartness field extension", and "solvable apartness field extension", it will be unpleasant.

**Comment by DanGrayson** [comment]
Re: "In the constructive setting this feeling gets a precise expression - the field of functions of a henzelization, which has decidable equality, is a field while Z_p is an object of a different class that is located slightly away from algebra in the direction of where the study of ordered algebraic structures lies."

Probably you're thinking of the fact that Z_p is a topological ring with a non-discrete topology.  But classically we may ignore the topology and consider it as a ring : its trivial ideal is prime, it's an integral domain, its fraction field is Q_p, and Q_p is a field.  That's all standard.  It would be a shame to throw all that away and relegate complete local rings and local fields to second class topics.

**Comment by DavidMichaelRoberts** [comment]
For a quickly readable reference, there's the nLab entry 'field'. It's a distillation of the various lengthy references given above, plus some extra insights from Toby Barrels. In particular it explains the relation between local rings and constructive notions of fields.
